packages/react-native/ReactCommon/react/nativemodule/idlecallbacks/NativeIdleCallbacks.cpp

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <react/renderer/runtimescheduler/RuntimeScheduler.h>
+#include <react/renderer/runtimescheduler/RuntimeSchedulerBinding.h>
+#include <chrono>
+#include "IdleTaskRef.h"
+#include "NativeIdleCallbacks.h"
+
+#ifdef RN_DISABLE_OSS_PLUGIN_HEADER
+#include "Plugins.h"
+#endif
+
+std::shared_ptr<facebook::react::TurboModule> NativeIdleCallbacksModuleProvider(
+    std::shared_ptr<facebook::react::CallInvoker> jsInvoker) {
+  return std::make_shared<facebook::react::NativeIdleCallbacks>(
+      std::move(jsInvoker));
+}
+
+namespace facebook::react {
+
+jsi::Function makeTimeRemainingFunction(
+    jsi::Runtime& runtime,
+    std::shared_ptr<RuntimeScheduler> runtimeScheduler,
+    RuntimeSchedulerTimePoint deadline) {
+  return jsi::Function::createFromHostFunction(
+    runtime,
+    jsi::PropNameID::forAscii(runtime, "timeRemaining"),
+    0,
+    [&](
+      jsi::Runtime& runtime,
+      const jsi::Value& /* unused */,
+      const jsi::Value* /* unused */,
+      size_t /* unused */) noexcept -> jsi::Value {
+        if (runtimeScheduler->getShouldYield()) {
+          return jsi::Value(runtime, 0.0);
+        }
+              
+        auto now = runtimeScheduler->now();
+        double remainingTime = std::chrono::duration_cast<std::chrono::milliseconds>(deadline - now).count();
+        return jsi::Value(runtime, std::max(remainingTime, 0.0));
+    });
+}
+
+NativeIdleCallbacks::NativeIdleCallbacks(std::shared_ptr<CallInvoker> jsInvoker)
+  : NativeIdleCallbacksCxxSpec(std::move(jsInvoker)) {}
+
+CallbackHandle NativeIdleCallbacks::requestIdleCallback(
+  jsi::Runtime& runtime,
+  jsi::Function callback,
+  std::optional<NativeRequestIdleCallbackOptions> options) {
+  using namespace std::chrono_literals;
+
+  auto binding = RuntimeSchedulerBinding::getBinding(runtime);
+  auto runtimeScheduler = binding->getRuntimeScheduler();
+
+  // handle timeout parameter
+  std::optional<double> timeout;
+  if (options.has_value()) {
+    timeout = options.value().timeout;
+  }
+
+  auto userCallback = std::make_shared<jsi::Function>(std::move(callback));
```

Comment:
This looks very sketchy. We shouldn't have shared pointers to `jsi::Function` because it prevents some of the safety mechanisms defined in that class. We should define the parameter in this function as `SyncCallback` instead, which allows us to move it into the lambda.

Reply:
I've just checked the code and it looks like I cannot use the `RawCallback&& variant`. This is because it uses the `std::function` under the hood, which is required to be copyable, so we cannot move capture the `callback` into the lambda.